source/extensions/filters/http/cdn/cdn_loop_parser.cc

```diff
@@ -0,0 +1,357 @@
+    break;
+  }
+
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "expected closing quote at position %d; found end-of-input", context.next()));
+  }
+
+  if (context.peek() != '"') {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "expected closing quote at position %d; found '%c'.", input.next(), context.peek()));
+  }
+  context.increment();
+
+  return context;
+}
+
+StatusOr<ParseContext> ParseToken(const ParseContext& input) {
+  ParseContext context = input;
+  while (!context.atEnd()) {
+    char c = context.peek();
+    if (c == '!' || c == '#' || c == '$' || c == '%' || c == '&' || c == '\'' || c == '*' ||
+        c == '+' || c == '-' || c == '.' || c == '^' || c == '_' || c == '`' || c == '|' ||
+        c == '~' || IsDigit(c) || IsAlpha(c)) {
+      context.increment();
+    } else {
+      break;
+    }
+  }
+  if (context.next() == input.next()) {
+    return absl::InvalidArgumentError(
+        absl::StrFormat("expected token starting at position %d.", input.next()));
+  }
+
+  return context;
+}
+
+StatusOr<ParseContext> ParsePlausibleIpV6(const ParseContext& input) {
+  ParseContext context = input;
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(absl::StrFormat(
+        "expected IPv6 literal at position %d; found end-of-input", context.next()));
+  }
+
+  if (context.peek() != '[') {
+    return absl::InvalidArgumentError(absl::StrFormat("expected opening '[' of IPv6 literal at "
+                                                      "position %d; found %c",
+                                                      context.next(), context.peek()));
+  }
+  context.increment();
+
+  while (true) {
+    if (context.atEnd()) {
+      break;
+    }
+    char c = context.peek();
+    if (!(IsHexDigitCaseInsensitive(c) || c == ':' || c == '.')) {
+      break;
+    }
+    context.increment();
+  }
+
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(
+        absl::StrFormat("expected closing ']' of IPv6 literal at position %d "
+                        "found end-of-input",
+                        context.next()));
+  }
+  if (context.peek() != ']') {
+    return absl::InvalidArgumentError(absl::StrFormat("expected closing ']' of IPv6 literal at "
+                                                      "position %d; found %c",
+                                                      context.next(), context.peek()));
+  }
+  context.increment();
+
+  return context;
+}
+
+StatusOr<ParsedCdnId> ParseCdnId(const ParseContext& input) {
+  ParseContext context = input;
+
+  if (context.atEnd()) {
+    return absl::InvalidArgumentError(
+        absl::StrFormat("expected cdn-id at position %d; found end-of-input", context.next()));
+  }
+
+  if (StatusOr<ParseContext> ipv6 = ParsePlausibleIpV6(context); ipv6) {
+    context.update(*ipv6);
+  } else {
+    if (StatusOr<ParseContext> token = ParseToken(context); !token) {
```

Comment:
it looks to me that uri-host contains more characters than token, so we may be unable to parse an otherwise valid cdn-id.

More specifically: reg-name    = *( unreserved / pct-encoded / sub-delims )

pct-encoded is not parsable. sub-delims contains more characters than token too.

Reply:
Valid pct-encoded is parsable by token.  RFC 3986 Appendix A says:
```
pct-encoded   = "%" HEXDIG HEXDIG
```
while token allows "%" and HEXDIG.  I'll add a test to make sure this sticks.

sub-delims is tricky because allowing all sub-delims makes the grammar ambiguous.  RFC 3986 Appendix A says:
```
sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="
```
In particular, `","` and `";"` make the grammar ambiguous since we don't know if the comma is part of the `uri-host` or if it delimits the start of the parameters (`";"`) or the start of the next entry.  You're right that token doesn't cover some of the other entries.  Let me dig into this a bit more to see if it makes sense to add a special parser to handle them.

Reply:
I read through the archives of the ietf-http-wg@w3.org mailing list where RFC 8586 was discussed (https://lists.w3.org/Archives/Public/ietf-http-wg/2018OctDec/thread.html and https://lists.w3.org/Archives/Public/ietf-http-wg/2019JanMar/thread.html for the curious, search for "cdn") to dig into the details a bit more.  Initially, the authors only wanted to allow tokens but were persuaded to allow hosts.

RFC 8586 says:
```
The cdn-id identifies the CDN using either a hostname under its control or a pseudonym.
```
The phrase "hostname under its control" corresponds to a `reg-name` in [RFC 3986 Section 3.2.2](https://tools.ietf.org/html/rfc3986#section-3.2.2]).  RFC 3986 says:
```
   A host identified by a registered name is a sequence of characters
   usually intended for lookup within a locally defined host or service
   name registry, though the URI's scheme-specific semantics may require
   that a specific registry (or fixed name table) be used instead.  The
   most common name registry mechanism is the Domain Name System (DNS).
   A registered name intended for lookup in the DNS uses the syntax
   defined in Section 3.5 of [RFC1034] and Section 2.1 of [RFC1123].
   Such a name consists of a sequence of domain labels separated by ".",
   each domain label starting and ending with an alphanumeric character
   and possibly also containing "-" characters.  The rightmost domain
   label of a fully qualified domain name in DNS may be followed by a
   single "." and should be if it is necessary to distinguish between
   the complete domain name and some local domain.

...

   This specification does not mandate a particular registered name
   lookup technology and therefore does not restrict the syntax of reg-
   name beyond what is necessary for interoperability.  Instead, it
   delegates the issue of registered name syntax conformance to the
   operating system of each application performing URI resolution, and
   that operating system decides what it will allow for the purpose of
   host identification.  A URI resolution implementation might use DNS,
   host tables, yellow pages, NetInfo, WINS, or any other system for
   lookup of registered names.  However, a globally scoped naming
   system, such as DNS fully qualified domain names, is necessary for
   URIs intended to have global scope.  URI producers should use names
   that conform to the DNS syntax, even when use of DNS is not
   immediately apparent, and should limit these names to no more than
   255 characters in length.
```

Given that Envoy assumes it's on the current Internet, we should assume DNS names.

The current implementation admits DNS names (and more), so I think it's appropriate here.

Do you want me to update the comment for `parseCdnId` with this discussion?

Reply:
Done.